src/backend/distributed/utils/metadata_cache.c

```diff
@@ -1772,3 +1836,64 @@ CitusInvalidateRelcacheByRelid(Oid relationId)
 		ReleaseSysCache(classTuple);
 	}
 }
+
+
+/*
+ * Register a relcache invalidation for the distributed relation associated
+ * with the shard.
+ */
+void
+CitusInvalidateRelcacheByShardId(int64 shardId)
+{
+	SysScanDesc scanDescriptor = NULL;
+	ScanKeyData scanKey[1];
+	int scanKeyCount = 1;
+	HeapTuple heapTuple = NULL;
+	Form_pg_dist_shard shardForm = NULL;
+	Relation pgDistShard = heap_open(DistShardRelationId(), AccessShareLock);
+
+	/*
+	 * Load shard, to find the associated relation id. Can't use
+	 * LoadShardInterval directly because that'd fail if the shard doesn't
+	 * exist anymore, which we can't have. Also lower overhead is desirable
+	 * here.
+	 */
+
+	ScanKeyInit(&scanKey[0], Anum_pg_dist_shard_shardid,
+				BTEqualStrategyNumber, F_INT8EQ, Int64GetDatum(shardId));
+
+	scanDescriptor = systable_beginscan(pgDistShard,
+										DistShardShardidIndexId(), true,
+										NULL, scanKeyCount, scanKey);
+
+	heapTuple = systable_getnext(scanDescriptor);
+	if (HeapTupleIsValid(heapTuple))
+	{
+		shardForm = (Form_pg_dist_shard) GETSTRUCT(heapTuple);
+		CitusInvalidateRelcacheByRelid(shardForm->logicalrelid);
+	}
+	else
+	{
+		/*
+		 * Couldn't find associated relation. That can primarily happen in two cases:
+		 *
+		 * 1) A placement row is inserted before the shard row. That's fine,
+		 *	  since we don't need invalidations via placements in that case.
+		 *
+		 * 2) The shard has been deleted, but some placements were
+		 *    unreachable, and the user is manually deleting the rows. Not
+		 *    much point in WARNING or ERRORing in that case either, there's
+		 *    nothing to invalidate.
+		 *
+		 * Hence we just emit a DEBUG5 message.
```

Comment:
Heh ;).  I wanted something that few of our tests turn on to catch pruning, because invalidations might  come from other backends, and that might not be stable.